<h1><center>TME 08 
    
    Features selection 
    Model selection 


# Import library

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import ElasticNet
from sklearn.svm import LinearSVC
from sklearn.feature_selection import SelectFromModel
from sklearn import linear_model
from sklearn.feature_selection import VarianceThreshold

# Import data

## Patients with leukemia

In [3]:
X1 = pd.read_csv('Golub_X',sep=' ',header = None) # Observations
y1 = pd.read_csv('Golub_y',sep=' ',header = None) # Classes
print(X1.shape)


(72, 3562)


## Breast cancer dataset

In [4]:
X = pd.read_csv('Breast.txt',sep=' ',header = None)
y2 = X.iloc[:,30] # Classes
X2 = X.iloc[:,0:29] # Observations
print(X2.shape)

(569, 29)


1. A simple heuristic approach is to delete features whose variance is less then a threshold. Try
it (with two different arbitrary thresholds) but do not expect this method to return an optimal
performance (although it can be quite efficient on some data sets).
http://scikit-learn.org/stable/modules/feature_selection.html

In [9]:
sel = VarianceThreshold(threshold=(0.05))
sel.fit_transform(X1)

print(sel.fit_transform(X1).shape)


(72, 387)


In [10]:
X1 = pd.read_csv('Golub_X',sep=' ',header = None) # Observations
y1 = pd.read_csv('Golub_y',sep=' ',header = None) # Classes
print(X1.shape)

sel = VarianceThreshold(threshold=(0.056))
print(sel.fit_transform(X1).shape)


(72, 3562)
(72, 143)


In [11]:
X = pd.read_csv('Breast.txt',sep=' ',header = None)
y2 = X.iloc[:,30] # Classes
X2 = X.iloc[:,0:29] # Observations
print(X2.shape)

sel = VarianceThreshold(threshold=(0.9))
sel.fit_transform(X2);
print(X2.shape)


(569, 29)
(569, 29)


This method isn't working for the second dataset !

In [12]:
y2 = X.iloc[:,30] # Classes
X2 = X.iloc[:,0:29] # Observations
print(X2.shape)
sel = VarianceThreshold(threshold=(0.001))
sel.fit_transform(X2);
print(X2.shape)

(569, 29)
(569, 29)


This method enable to filter variables that have a very poor variance. By default it removes variance equal to 0. We can add a threshold.

We can alos use statistical test to filter our data. The filtred data will be the one that are significant according to the choosed test, here alpha is the threshold at 5%. We can use a chi2, but only with matrix with non-negatives values

2. Univariate feature selection with statistical tests to get rid of features which are not statistically significant with respect to the vector of class. Try the SelectFdr function that computes
p-values for an estimated false discovery rate.
http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFdr.
html#sklearn.feature_selection.SelectFdr

In [13]:
from sklearn.feature_selection import SelectFdr, chi2

In [14]:

X_new = SelectFdr(alpha=0.05).fit_transform(X2, y2)
X_new.shape

(569, 24)

Here, 3 colomns where filtred for X2

In [15]:
X_new = SelectFdr(alpha=0.05).fit_transform(X1, np.ravel(y1))
X_new.shape

(72, 545)

Here, we have more than 3000 colomns that where filtred

3. L1-based feature selection is designed to find an optimal solution. The sparsity parameter
is important (since it controls the number of non-zero parameters: it too many parameters
are kept, no really feature selection; if too few parameters are chosen, it is possible that the
accuracy is very poor).

    (a) Logistic regression penalized by the L1 penalty term 
    
    $linear_model.Lasso(alpha=alpha)$
    
    (b) A support vector machine penalized by the L1 penalty term
    
$LinearSVC(C=C, penalty="l1", dual=False)$

    (c) Explore the Elastic Net which is a compromise between the L1 and L2 penalty terms.
    
$ElasticNet(alpha=alpha, l1_ratio=0.7)$

In [33]:
X1 = pd.read_csv('Golub_X',sep=' ',header = None) # Observations
y1 = pd.read_csv('Golub_y',sep=' ',header = None) # Classes
print(X1.shape)

a = linear_model.Lasso(alpha=0.05)
f1 = a.fit(X1,y1)
model = SelectFromModel(f1,prefit=True)
Xnew = model.transform(X1)
Xnew.shape


(72, 3562)


(72, 4)

In [25]:
X1 = pd.read_csv('Golub_X',sep=' ',header = None) # Observations
y1 = pd.read_csv('Golub_y',sep=' ',header = None) # Classes
print(X1.shape)
b = LinearSVC(C = 0.05,penalty="l1", dual=False)
f2 = b.fit(X1,np.ravel(y1))
m2 = SelectFromModel(f2,prefit=True)
Xnew = model.transform(X1)
Xnew.shape

(72, 3562)


(72, 38)

In [32]:
X1 = pd.read_csv('Golub_X',sep=' ',header = None) # Observations
y1 = pd.read_csv('Golub_y',sep=' ',header = None) # Classes
print(X1.shape)

c = ElasticNet(alpha=0.05, l1_ratio=0.7,copy_X=False)
f3 = c.fit(X1,y1)
m2 = SelectFromModel(f3,prefit=True)
Xnew = m2.transform(X1)
Xnew.shape
print(Xnew.shape)

(72, 3562)
(72, 14)


### for X2 data

In [34]:
y2 = X.iloc[:,30] # Classes
X2 = X.iloc[:,0:29] # Observations
print(X2.shape)

(569, 29)


In [35]:
y2 = X.as_matrix()[:,30] # Classes
X2 = X.as_matrix()[:,0:29] # Observations
print(X2.shape)

b = linear_model.Lasso(alpha=0.01,copy_X=False)
f4 = b.fit(X2,y2)

m2 = SelectFromModel(f4,prefit=True)
Xnew = m2.transform(X2)
Xnew.shape

(569, 29)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


(569, 16)

In [36]:
y2 = X.as_matrix()[:,30] # Classes
X2 = X.as_matrix()[:,0:29] # Observations
print(X2.shape)

b = LinearSVC(C = 0.05,penalty="l1", dual=False)
f6 = b.fit(X2,y2)
m2 = SelectFromModel(f6,prefit=True)
Xnew = m2.transform(X2)
Xnew.shape

(569, 29)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


(569, 10)

In [37]:
y2 = X.as_matrix()[:,30] # Classes
X2 = X.as_matrix()[:,0:29] # Observations
print(X2.shape)

c = ElasticNet(alpha=0.05, l1_ratio=0.7)
f8 = c.fit(X2,y2)
m2 = SelectFromModel(f8,prefit=True)
Xnew = m2.transform(X2)
Xnew.shape

(569, 29)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


(569, 6)

4. How many features do you keep using these different methods? It is quite normal that each
method selects a different number of features.

All these methods leads to differents results. As you have a huge matrix to analyze, it's crucial to filter our data first. The fist methods using the threshold reduce our matrix from 3562 to 387 for X1 with alpha = 0.05, and to 143 for alpha = 0.056. For X2 with very high or very slow threshold, it dosen't change anything
For the second methods the matrix was reduced of 3 columns for X1 and ofmore than 3000 for X2 with alpha = 0.05.
Finaly, the L1 feature selection give matrix of 4 rows for X1 with Lasso method (alpha = 0.05), the SVM methods gives 38 (alpha = 0.05) and the Elastic net methods gives 14 for X1. 
For X2 respectively we have now 14 columns 10 columns and 6 columns for these 3 methods.

5. What method leads to the best performance (on the given data sets) ?

If we suppose that the best methods is the ones that filtred more variables (whitch is not always the case, the .  methods works better.
The Elastic net methods is supposed to be the better because combine L1 and L2 linear regression penality terms.
According to our data, it could be too restrivctive for alpha = 0.05, but as we don't know the meaning of our data, it's complicated to conclude which methods is the better